Expose cycle counter and use it in /channelz.

soheilhy opened the pull request:
We are using clock real time (which is very expensive) on every
RPC to update channelz. We should simply use cycle clock.

This patch exposes cycle clock, enables RDTSC on intel,
and use it for channelz.

This is part of a larger performance series.

Sorry I have to bother many of you:
@nicolasnoble  for portability
@AspirinSJL for time
@yashykt  for /channelz

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/grpc/grpc/18937)
<!-- Reviewable:end -->

soheilhy commented:
The bazel failure are because of the test doesn't support precise clock (assumes precise clock is always realtime clock /facepalm). 

Fixing it ...

soheilhy commented:
Thank you for the review @AspirinSJL 

I'll address them as soon as I figure out how I can fix the broken test. The test is still broken, so please don't review until I handle that. I think there is an unrelated bug that resurfaces here.

AspirinSJL commented:
The bug at src/core/lib/gpr/time_precise.cc, line 71 seems real. Maybe try fixing that first?

soheilhy commented:
> The bug at src/core/lib/gpr/time_precise.cc, line 71 seems real. Maybe try fixing that first?

Thank you for catching that, but no that's definitely not the issue. :-)

Note that this is a new API so existing and unmodified tests don't use it. On top of that, the test breaks when RDTSC is enabled, and that code path is eliminated by the macro, and passes when the macro is undefined. It's something else, but I don't still know what it is.

soheilhy commented:
Figured it out. The reason is that it fails is that `gpr_precise_clock_init()` loops for 10s to find the frequency, and that creates a 10s gap from the time origin of the cycle clock which is used for the deadlines in the iomgr.  The test uses 5s timeouts, and that doesn't seem to play well with that 10s of gap.

Lowering the spin time of  gpr_precise_clock_init() down to 1s, or in fixes the issue, but that's a hack I don't like, because I think something else is broken and this is just a sympom. I'll see which one is going to turn out working best tomorrow when I have more caffeine in my blood.

soheilhy commented:
Once I figure why the deadline fires due to the 10s loop, and fix the root cause, I will change this patch to avoid looping for 10s. Abseil uses 64ms and that should be enough:
https://github.com/abseil/abseil-cpp/blob/master/absl/base/internal/sysinfo.cc#L180-L212

We can really afford delaying every grpc server start time by 10s on Intel. :-)

soheilhy commented:
> gpr_precise_clock_init() loops for 10s

This is actually the root cause. It results in the test server taking 10s to bootstrap causing all client-side deadlines to expire. Fix is what I said last night, to replicate what abseil does a shorter/smarter loop and potentially reading the tsc_freq file when it's exported.

soheilhy commented:
The patch is working as intended. Essentially all end2end benchmarks show 19% to 45% now call reduction.  :-)

soheilhy commented:
Failures are #18967 and #18511

soheilhy commented:
@markdroth  I'm going to send the inlinedvector changes as a separate PR. Many tests started failing as a result of that change, so it might be worth having that separately.

soheilhy commented:
I think I fixed the issue. One more try ...

soheilhy commented:
yay! All tests pass ... :-)

yashykt commented:
I'm deferring to Mark's review on channelz :)

soheilhy commented:
@nicolasnoble friendly ping. This is pending for ~2 months now.

I disabled it by default for open source, in case that helps expediting the review.

hcaseyal commented:
@hcaseyal cc'ing myself on this

soheilhy commented:
Thank you @hcaseyal for the follow up and thank you @nicolasnoble for the approval! 

Just squashed and rebased the patch series.  Will wait for the tests to pass.

soheilhy commented:
Known issue #19819
